### PR TITLE
add AVX2 implementation for sigmoid function

### PR DESCRIPTION
**Pull request by vedanuj**
PR introduces AVX2 optimization for sigmoid floats. Issue #4929. The internal benchmark shows ~10x speedup.

- Added AVX2 vectorized sigmoid using the 8-way vectorized `exp` (`exp256_ps`) in _avx_mathfun.h_. 

- Implemented vector dispatch for sigmoid. Since sigmoid function is defined for floats and doubles only, for now, added preprocessor `#ifdef` to init sigmoid dispatch only for float and double.

- Vector functions in _THVector.h_ were not called for all of the basic functions in floating point or double only. Changed the `LAB_IMPLEMENT_BASIC_FUNCTION` define in _THTensorMatch.c_ to use `THVector_(NAME)` implementations if the inputs are contiguous. For the functions that do not have vectorized SIMD implementations will use the same default function from _THMath.h_

**Benchmark**

Non-vectorized sigmoid : 
```
In [1]: import torch
In [2]: x = torch.randn(10000,10000)
In [3]: %time _ = x.sigmoid()
CPU times: user 2.8 s, sys: 130 ms, total: 2.93 s
Wall time: 737 ms
```
```
In [1]: import torch
In [2]: x = torch.randn(1000,1000)
In [3]: %time _ = x.sigmoid()
CPU times: user 29.1 ms, sys: 4.16 ms, total: 33.3 ms
Wall time: 8.63 ms
```

AVX2 Vectorized sigmoid

```
In [1]: import torch
In [2]: x = torch.randn(10000,10000)
In [3]: %time _ = x.sigmoid()
CPU times: user 206 ms, sys: 106 ms, total: 312 ms
Wall time: 78.2 ms
```

```
In [14]: x = torch.randn(1000,1000)
In [15]: %time _ = x.sigmoid()
CPU times: user 179 µs, sys: 2.95 ms, total: 3.13 ms
Wall time: 858 µs
```

### CODE REVIEW

**Comment by vedanuj** [comment]
PR for Issue #4929

File: aten/src/TH/vector/AVX2.cpp [hunk 1]
```diff
@@ -105,18 +105,25 @@ void THFloatVector_normal_fill_AVX2(float *data,
...
-    YMM0 = _mm256_loadu_ps(x+i);
-    YMM1 = _mm256_add_ps(one, exp256_ps(YMM0));
-    YMM2 = _mm256_div_ps(one, YMM1);
-    _mm256_storeu_ps(y+i, YMM2);
+  const __m256 minus_one = _mm256_set1_ps(-1.0f);
+  __m256 YMM0, YMM1, YMM2, YMM3;
+  for (i = 0; i <= ((n)-16); i += 16) {
+    YMM0 = _mm256_loadu_ps(x + i);
+    YMM1 = _mm256_loadu_ps(x + i + 8);
+    YMM0 = _mm256_mul_ps(minus_one, YMM0);
```
Comment (zdevito) [inline aten/src/TH/vector/AVX2.cpp]: Is `_mm256_sub_ps(zero, YMM0)` or an xor of the floating point sign bit cheaper than a multiply here? Though it is probably not that important since exp256 dominates.

File: aten/src/TH/generic/THTensorMath.c [hunk 2]
```diff
@@ -3629,22 +3629,32 @@ TENSOR_IMPLEMENT_LOGICAL(ne,!=)
...
-    }                                                                                                        \
-    else {                                                                                                   \
-      TH_TENSOR_APPLY2(real, r_, real, t, *r__data = CFUNC(*t_data););                                       \
-    }                                                                                                        \
+    if (r_Contig && tContig) {                                \
+      TH_TENSOR_APPLY2_CONTIG(real, r_, real, t, THVector_(NAME)(r__data, t_data, r__len););                   \
+    } else {                                                                                                   \
+      int inOMP = omp_in_parallel();                            \
+      if( (r_Size > TH_OMP_OVERHEAD_THRESHOLD) && (!inOMP) ){   \
+        TH_TENSOR_APPLY2_OMP(r_Size, r_Contig, tContig, real, r_, real, t, *r__data = CFUNC(*t_data););        \
```
Comment (zdevito) [inline aten/src/TH/generic/THTensorMath.c]: I am concerned about modifying this macro for all the non-AVX functions just so that sigmoid gets remapped. Is it possible to write this so _just_ the functions we want (for now, just sigmoid) check and dispatch to their vector equivalent?  @gchanan do you know why TH_TENSOR_APPLY2_OMP no longer uses TH_TENSOR_APPLY2_CONTIG by default?

**Review by zdevito — COMMENTED** [review]
This looks good! Let's see if we can track down what is going on with THVector_ dispatch to make sure we aren't degrading performance in some way for the non-vectorized functions by adding the THVector_ dispatch.

File: aten/src/TH/generic/THTensorMath.c (hunk 2 above)
Comment (fmassa) [reply aten/src/TH/generic/THTensorMath.c]: @zdevito `TH_TENSOR_APPLY2_OMP` was recently introduced and is a superset of `TH_TENSOR_APPLY2_CONTIG`, and handles the case where the inputs are non-contiguous as well.
I believe we can now remove `TH_TENSOR_APPLY2_CONTIG` as this case is covered by the `*_OMP` macros.
Also, as it fuses the contiguous dimensions, the overhead of using it compared to `*CONTIG` should be minimal I think.

But, if we remove the `*CONTIG` macros, I think we would need to pay attention to the fact that 
[`TH_TENSOR_APPLY2_OMP` tries to force vectorization at the compiler level for contiguous tensors](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/THTensorApply.h#L372-L388), which might make the binaries that we release not work on cpus without vector instructions, as it would lack the runtime dispatch.

Comment (soumith) [reply aten/src/TH/generic/THTensorMath.c]: > TH_TENSOR_APPLY2_OMP tries to force vectorization at the compiler level for contiguous tensors

This does not matter. We set AVX/AVX2 flags only for particular C files. So even if compiler tries to vectorize other files, it wont use AVX/AVX2 intrinsics.

Comment (zdevito) [reply aten/src/TH/generic/THTensorMath.c]: The code as written seems to _disable_ vectorization by bypassing THVector_ methods, which is why @vedanuj needs to route around it to get sigmoid to vectorize. That seems bad when we have a custom vectorized implementation, since we will fallback to a worse version. @fmassa Do you have a idea of how to fix these cases so that they can work together? Ideally we fallback to trying to force the compiler to vectorize only when we don't already have a custom vectorized implementation.  The compiler is not going to be able to vectorize sigmoid since it doesn't have a vectorized exp.

Comment (vedanuj) [reply aten/src/TH/generic/THTensorMath.c]: @zdevito I can make these changes into a new macro for vectorized functions(for now only sigmoid will use that) and any other functions that will be vectorized later can use this new macro. However changing the LAB_IMPLEMENT_BASIC_FUCNTION macro doesn't hurt  perf for the other non vectorized functions because their THVector implementations are calling the serialized version of the functions in [` VECTOR_IMPLEMENT_FUNCTION`](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THVectorDefault.cpp#L185). I tested a few functions like tanh, log, exp and didn't notice any degradation in their perf with this change.

File: aten/src/TH/vector/AVX2.cpp (hunk 1 above)
Comment (vedanuj) [reply aten/src/TH/vector/AVX2.cpp]: Tried with `_mm256_sub_ps(zero, YMM0)`. Doesn't give any noticeable wins as exponent is the more expensive.

**Comment by vedanuj** [comment]
### Changes in new commit : 

- Added a new macro `LAB_IMPLEMENT_VECTORIZED_FUNCTION` for the vectorized basic functions. Currently only `sigmoid` uses this macro which redirects to the vectorized implementation.

- Although there is no significant improvement in performance(due to exponent being the computationally dominant operation),  replaced `_mm256_mul_ps(minus_one, YMM0)` with `_mm256_sub_ps(zero, YMM0)` which should be computationally less expensive.

(Please re-review the code @zdevito @soumith  @fmassa)

**Comment by goldsborough** [comment]
@cpuhrsch do you want to take a look at this? I know you're working on CPU improvements.

**Comment by cpuhrsch** [comment]
@goldsborough I'm writing those directly within ATen native, so this code won't conflict with anything I'm writing right now.

**Comment by ezyang** [comment]
There is probably more to look into re this code path, but I don't see why we shouldn't take such an obvious improvement for sigmoid.
